File: Python_Codes/6. Collections Framework/Collections/ComparableExample.py
from dataclasses import dataclass

@dataclass(order=True)
class Student:
    grade: int
    name: str  # Note: grade comes first to make sorting by grade the default

    def __str__(self):
        return f"Student(name='{self.name}', grade={self.grade})"


def main():
    students = [
        Student(name="Alice", grade=85),
        Student(name="Bob", grade=92),
        Student(name="Charlie", grade=78)
    ]

    print("Before sorting:")
    for student in students:
        print(student)

    # Sorts by grade because of the order of fields in the dataclass
    students.sort()

    print("\nAfter sorting:")
    for student in students:
        print(student)


if __name__ == "__main__":
    main()

File: Java_Codes/6. Collections Framework/ComparableExample.java
package Collections;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Student implements Comparable<Student> {
    private String name;
    private int grade;

    public Student(String name, int grade) {
        this.name = name;
        this.grade = grade;
    }

    public String getName() {
        return name;
    }

    public int getGrade() {
        return grade;
    }

    @Override
    public int compareTo(Student other) {
        return Integer.compare(this.grade, other.grade);
    }

    @Override
    public String toString() {
        return "Student{name='" + name + "', grade=" + grade + "}";
    }
}

public class ComparableExample {
    public static void main(String[] args) {
        List<Student> students = new ArrayList<>();
        students.add(new Student("Alice", 85));
        students.add(new Student("Bob", 92));
        students.add(new Student("Charlie", 78));

        System.out.println("Before sorting:");
        for (Student student : students) {
            System.out.println(student);
        }

        Collections.sort(students);

        System.out.println("\nAfter sorting:");
        for (Student student : students) {
            System.out.println(student);
        }
    }
}
